### PR TITLE
Fix failure with missing output file for UnexportMainSymbol

### PR DESCRIPTION
**Pull request by big-guy**
Fixes https://github.com/gradle/gradle-native/issues/304

### Contributor Checklist
- [ ] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/.github/CONTRIBUTING.md)
- [ ] [Sign Gradle CLA](http://gradle.org/contributor-license-agreement/)
- [ ] Check ["Allow edit from maintainers" option](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) in pull request so that additional changes can be pushed by Gradle team
- [ ] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective
- [ ] Provide unit tests (under `<subproject>/src/test`) to verify logic
- [ ] Update User Guide, DSL Reference, and Javadoc for public-facing changes
- [ ] Ensure that tests pass locally: `./gradlew <changed-subproject>:check`

### Gradle Core Team Checklist
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation
- [ ] Recognize contributor in release notes

### CODE REVIEW

File: subprojects/language-native/src/integTest/groovy/org/gradle/language/swift/tasks/UnexportMainSymbolIntegrationTest.groovy [hunk 1]
```diff
@@ -35,10 +38,32 @@ class UnexportMainSymbolIntegrationTest extends AbstractInstalledToolChainIntegr
...
+    def "clean build works"() {
+        writeMainSwift()
+
+        expect:
+        succeeds("clean", "unexport", "assemble")
+    }
+
+    @NotYetImplemented
+    @Issue("https://github.com/gradle/gradle-native/issues/297")
+    def "unexport is incremental"() {
```
Comment (lacasseio) [inline subprojects/language-native/src/integTest/groovy/org/gradle/language/swift/tasks/UnexportMainSymbolIntegrationTest.groovy]: Should we make a separate PR for this fix? I don't think we should merge code that adds NotYetImplmented. What do you think?

File: subprojects/language-native/src/main/java/org/gradle/language/swift/tasks/UnexportMainSymbol.java [hunk 2]
```diff
@@ -71,7 +71,8 @@ public File getMainObject() {
     public void unexport() {
         final File mainObjectFile = getMainObject();
         if (mainObjectFile != null) {
-            final File relocatedMainObject = outputFile.file(mainObjectFile.getName()).get().getAsFile();
+            final File relocatedMainObject = outputDirectory.file(mainObjectFile.getName()).get().getAsFile();
+            relocatedMainObject.getParentFile().mkdirs();
```
Comment (lacasseio) [inline subprojects/language-native/src/main/java/org/gradle/language/swift/tasks/UnexportMainSymbol.java]: Could you have a getter for the directory and annotate it with @OutputDirectory so you don't have to mkdirs here? Does that even make sense? Probably what you have there make senses.

File: subprojects/language-native/src/integTest/groovy/org/gradle/language/swift/tasks/UnexportMainSymbolIntegrationTest.groovy (hunk 1 above)
Comment (big-guy) [reply subprojects/language-native/src/integTest/groovy/org/gradle/language/swift/tasks/UnexportMainSymbolIntegrationTest.groovy]: I thought it might have been fixed by the same change, but it wasn't. There's something deeper/weirder going on.  This looked like a bug we would fix soon-ish anyways, so I thought it was worth keeping the test.

File: subprojects/language-native/src/main/java/org/gradle/language/swift/tasks/UnexportMainSymbol.java (hunk 2 above)
Comment (big-guy) [reply subprojects/language-native/src/main/java/org/gradle/language/swift/tasks/UnexportMainSymbol.java]: Yeah, there are two things going on:

- To link this object optionally (when there are no sources), we need a `FileTree`
- To get the directory created automatically, we need a `@OutputDirectory`.

We could do something like:
```
@Internal FileCollection getObjects()
@OutputDirectory Provider<Directory> getOutputDirectory()
```

I can try that.

File: subprojects/language-native/src/integTest/groovy/org/gradle/language/swift/tasks/UnexportMainSymbolIntegrationTest.groovy [hunk 3]
```diff
@@ -35,10 +37,31 @@ class UnexportMainSymbolIntegrationTest extends AbstractInstalledToolChainIntegr
...
+    @Issue("https://github.com/gradle/gradle-native/issues/304")
+    def "clean build works"() {
+        writeMainSwift()
+
+        expect:
+        succeeds("clean", "unexport", "assemble")
+    }
+
+    @Issue("https://github.com/gradle/gradle-native/issues/297")
+    def "unexport is incremental"() {
```
Comment (big-guy) [inline subprojects/language-native/src/integTest/groovy/org/gradle/language/swift/tasks/UnexportMainSymbolIntegrationTest.groovy]: OK, and now it does actually fix this.

File: subprojects/language-native/src/integTest/groovy/org/gradle/language/swift/tasks/UnexportMainSymbolIntegrationTest.groovy [hunk 4]
```diff
@@ -35,10 +37,31 @@ class UnexportMainSymbolIntegrationTest extends AbstractInstalledToolChainIntegr
         """
     }
 
+    @Issue("https://github.com/gradle/gradle-native/issues/304")
+    def "clean build works"() {
+        writeMainSwift()
+
+        expect:
+        succeeds("clean", "unexport", "assemble")
```
Comment (lacasseio) [inline subprojects/language-native/src/integTest/groovy/org/gradle/language/swift/tasks/UnexportMainSymbolIntegrationTest.groovy]: Why do you need to call `unexport`? Should that task be implementation detail and doesn't expect a user to call that task directly. I was able to get the same behavior with `succeeds("clean", "test")`. I don't know if that has the same intent as this test. What do you think?

File: subprojects/language-native/src/integTest/groovy/org/gradle/language/swift/tasks/UnexportMainSymbolIntegrationTest.groovy (hunk 3 above)
Comment (lacasseio) [reply subprojects/language-native/src/integTest/groovy/org/gradle/language/swift/tasks/UnexportMainSymbolIntegrationTest.groovy]: W00t!

File: subprojects/language-native/src/integTest/groovy/org/gradle/language/swift/tasks/UnexportMainSymbolIntegrationTest.groovy (hunk 4 above)
Comment (big-guy) [reply subprojects/language-native/src/integTest/groovy/org/gradle/language/swift/tasks/UnexportMainSymbolIntegrationTest.groovy]: This is a simplified version of what the test plugin does (the build script directly creates the `UnexportMainSymbol` task).

We're calling `unexport` because there's nothing actually using the output of `unexport`.  We're just looking at the generated object file.  There are some tests in `SwiftXCTestIngrationTest` that uses the test plugin and `UnexportMainSymbol` task.

The calls to `assemble` are actually misleading in the other tests.  There used to be something more complicated here that checked that the final executable had a valid `_main` symbol.  I went ahead and removed them.
